### ISSUE TITLE
KeyError: device_times and more

### ISSUE THREAD

**Issue by GeoffMorg** [markgdev/home-assistant_OctopusAgile#27]
I am a very inexperienced HA user (but have a fairly strong computer tech background). HA is running on a Pi 3 and it primarily for use with Agile so OctopusAgile is a "must-have"!!

I have installed it according to the .md file and copy/pasted your example config.yaml entry, and added my MPAN, serial and AUTH. Other than that I don't think I've made any changes,

I was expecting to see a bunch of new sensors in my entities but I have just 4 - rates, region code, startdate and timers. In particular your example has dishwasher, washer, wash_dry, switch.water_heater. Maybe I need to do something else to create these but I thought they came about from the config.yaml entries.

Anyway I checked the log and I have an error shortly after startup. It looks like this

2020-09-01 18:14:17 ERROR (MainThread) [homeassistant.setup] Error during setup of component OctopusAgile
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/setup.py", line 191, in _async_setup_component
    result = await task
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/OctopusAgile/__init__.py", line 72, in setup
    for entity_id, vals in data["device_times"].items():
KeyError: 'device_times'

My advance apologies if I am being a real muppet but I'd appreciate some help in resolving this. I'd have sooner not raised it as an issue but I can't see any other way to contact you.

Thanks

**Comment by markgdev** [comment]
I think the /config/octopusagile.json file which stores the sensor info has got something bad in it/missing something. If you delete that and restart, I think your issue may resolve itself with a bit of luck. Let me know if that sorts it out.

Mark

**Comment by GeoffMorg** [comment]
Thanks for the prompt response. I deleted that file but it hasn't recreated it. I also noticed a message in the log which I'm pretty sure wasn't there before - "Unable to find services.yaml for the octopusagile integration" but I'm not seeing the error I previously reported.

**Comment by markgdev** [comment]
The services.yaml error is expect, it’s because I haven’t added any helper info to the service yet.

If you manually call the octopusagile.half_hour and octopusagile.update_timers services, do you get any errors/does anything update?

**Comment by GeoffMorg** [comment]
Well... I restarted the Pi and when I came back much later there were about 15 OctopusAgile entities, including the dishwasher and others I expected. So I rebooted again and after boot it was back to 5. So I did the testing for your questions above and some reading, maybe 10 minutes, and looked again and there are now more entities. I think they could just be really REALLY slow appearing after a boot (though I'm pretty sure they weren't originally coming in so maybe that file deletion helped).!
Calling octopusagile.half_hour I get an on-screen error error "Failed to call service octopusagile/half_hour. string indices must be integers"
Calling octopusagile.update_timers I get an on-screen error error "Failed to call service octopusagile/half_hour. min() arg us an empty sequence"
However I'm doing this from within Developer/Service Tools and maybe they both need me to type in some service data?
Thanks again for your help!

**Comment by markgdev** [comment]
@GeoffMorg, sorry for the delay. Did you get any further with this? I’ve had someone else but this issue but I wasn’t able to replicate. I have a feeling that things sorted themselves out after home assistant had been up for a little while and all the entities had updated themselves.
ref the entities being slow to show up after boot, the way I have it setup is to update at the next half hour point. I should be able to change this so will add it to the list for when I get around to spending some more time on it.

**Comment by GeoffMorg** [comment]
Thanks for coming back to me!! :-) I've not spent a lot of time with this over the past week or so but today I've made an effort to get it going, with quite some success.

Firstly I was trying to run before I can walk. I'd tried putting the config in an include file rather than filling up config.yaml, and although it reported no errors this made it not work. I need to revisit my understanding of how to do that!!

 I had also changed some names and values to ones I would need and I think I'd ended up with some badly spaced some items.

Anyway, I've now re-installed everything and deleted the json file etc and am now getting most of the sensors eventually  but I think I am missing most of the timers simply because they don't populate for ages after startup - and because I am still testing I am rebooting too frequently to ever see them. For instance, having just rebooted I am getting update_timers_nextupdate is showing as 19:00Z which is still nearly 2 hours away, and I am not seeing any of the timers even exist,

I think if you were able to set it (or advise me how I can set it temporarily) to update once it is turned on rather than having to wait it would help me immensely.  (I'm going to have a poke about in the code to see if I can work it out but I'd prefer you advised!!)

Thanks again.
Geoff

**Comment by markgdev** [comment]
Aha, yes, I failed with telling you to only call octopusagile.half_hour. 

What you actually need to do is call octopusagile.update_timers and then call octopusagile.half_hour. I think calling half_hour first is what's corrupting the Json file. Once the services have run correctly the json file should repopulate most entities after a reboot (if I remember rightly. I don't tend to restart that much now I'm all setup.)

**Comment by GeoffMorg** [comment]
I really appreciate the quick response!

Yeah, my HA is up and down like the proverbial, cos I'm actively setting quite a bit of stuff up!!

I am n pw calling these through Developer Tools/Services, is that where you mean? I'm not adding any Service Data.

octopusagile.update_timers services briefly display a black message box across the bottom of the screen "Failed to call octopusagile.update_timers services. 'float'object cannot be interpreted as integer."

I guess octopusagile.half_hour is failing because of the above but I get "Failed to calloctopusagile.half_hour. expected dict for dictionary value @ data['service data']: Got 30"

PS I looked at the code and decided I will leave any changes there to you!! :-)

**Comment by markgdev** [comment]
Yea through the developer tab, no service data needed.
Do you happen to have a decimal value for numhrs in one of your timers? I'm wondering if that is the case whether I've not handled it correctly. I only have complete hours in my config, but I should've handled half hours if I haven't!
If it's not that, happy to see if I can see anything in your config if you want to send it over.

**Comment by GeoffMorg** [comment]
You mean in the run devices? I do, but I think your dishwasher example did too.

 

Anyway I’ve taken out a couple of 0.5 in the timers bit. And that service now runs.

 

I could do with the 0.5 working of that turns out to be it, as I’m hoping to just have the car charging on the cheapest half hours rather than a block, but I could live without…

 

And the update timers nextupdate is looking healthier now it’s run.

 

I can’t do it now because some thirsty neighbours just turned up but later I will mess around and see what I can find.

 

Progress!! :)

 

Thanks

**Comment by markgdev** [comment]
The timers will always give you the cheapest half hours and not in a block as it currently works. Numhrs is just the total number of hours you need. It would be perfectly acceptable to only want half an hour though. Run devices will give you a block. I do plan to add the ability into timers for blocks too though.
I will add in a half hour timer to my config tomorrow and try and track down the issue.

**Comment by markgdev** [comment]
I've just pushed up a fix for the decimal number of hours issue. I managed to replicate it straight away. Funny enough I wanted to add my towel rail to come on for 1.5 hours and had been meaning to add it into my config so would've come across it myself in a few days anyway!

**Comment by GeoffMorg** [comment]
Thanks Mark, been testing with the new version today. There is something a bit weird going on but it may well be due to something I've done. I put it all pretty much back to the example above and one time I definitely saw some times against the dishwasher etc entities , but now they report Entity not available. I haven't delved into that too much as I think there us a more fundamental problem which could resolve it. That is, the timers are not getting populated - calling the service manually gives me 
"Failed to call service octopusagile/update_timers.min{}arg is an empty sequence" and I can see this in the Core log:

 Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/components/websocket_api/commands.py", line 125, in handle_call_service
    await hass.services.async_call(
  File "/usr/src/homeassistant/homeassistant/core.py", line 1308, in async_call
    task.result()
  File "/usr/src/homeassistant/homeassistant/core.py", line 1347, in _execute_service
    await self._hass.async_add_executor_job(handler.func, service_call)
  File "/usr/local/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/OctopusAgile/__init__.py", line 195, in handle_half_hour_timer
    timers = hass.states.get("octopusagile.timers").attributes["timers"]
AttributeError: 'NoneType' object has no attribute 'attributes'

Would it help if I posted up the relevant bits of my config.yaml? Or sent it on email? Not sure if you are on the Octopus Agile forum but I could PM you there, which would clutter up this page less. I am on there as Geoff.

btw one thing which def helps sometimes is deleting that .json file, but it doesn't always exist.


Cheers
Geoff

**Comment by GeoffMorg** [comment]
Uh-oh, I think I have misled myself in the latest post.  I did say I'm a novice! :-)

In trying to clarify for myself what was what in my config.yaml and why I wasn't seeing stuff in my Lovelace I had given the entities a different but similar prefix and I had the wrong names in the card on Lovelace. I think it's all looking OK now and I can crack on with some testing.
I will go back to implementing the half hours - but just for clarity, if I want to charge the car on the best 5 half hour sessions in a night on a switch called car_switch, I'd put that as a timer as follows:
  timers:
  - day_from: tomorrow
    day_to: tomorrow
    entity_id: switch.car_switch
    numHrs: 2.5
    params: null
    requirements:
      - day_from: tomorrow
        day_to: tomorrow
        numHrs: 2.5
        time_from: '00:00:01'
        time_to: '06:00:00'
   
(I'll then just need to work out how to get a variable to numHrs which will depend on how empty the car is.)
